internal/cache/file/cache_handler.go

```diff
@@ -0,0 +1,224 @@
+	mu locker.Locker
+}
+
+func NewCacheHandler(fileInfoCache *lru.Cache, jobManager *downloader.JobManager, cacheLocation string, filePerm os.FileMode) *CacheHandler {
+	return &CacheHandler{
+		fileInfoCache: fileInfoCache,
+		jobManager:    jobManager,
+		cacheLocation: cacheLocation,
+		filePerm:      filePerm,
+		mu:            locker.New("FileCacheHandler", func() {}),
+	}
+}
+
+func (chr *CacheHandler) createLocalFileReadHandle(objectName string, bucketName string) (*os.File, error) {
+	fileSpec := data.FileSpec{
+		Path: util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(bucketName, objectName)),
+		Perm: chr.filePerm,
+	}
+
+	return util.CreateFile(fileSpec, os.O_RDONLY)
+}
+
+func (chr *CacheHandler) cleanUpEvictedFile(fileInfo *data.FileInfo) error {
+	key := fileInfo.Key
+	_, err := key.Key()
+	if err != nil {
+		return fmt.Errorf("error while performing post eviction: %v", err)
+	}
+
+	// Removing Job doesn't delete the job object itself but invalidates the job and
+	// hence it's possible that some existing cache handle will have reference to older
+	// job object, but that job object will have INVALID status.
+	chr.jobManager.RemoveJob(key.ObjectName, key.BucketName)
+
+	localFilePath := util.GetDownloadPath(chr.cacheLocation, util.GetObjectPath(key.BucketName, key.ObjectName))
+	err = os.Remove(localFilePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logger.Warnf("while clean up, filePath should exist: %v", err)
+		} else {
+			return fmt.Errorf("while deleting file: %s, error: %v", localFilePath, err)
+		}
+	}
+
+	return nil
+}
+
+// addFileInfoEntryToCache adds a data.FileInfo entry for the given object and bucket
+// in the file info cache if it does not already exist. While adding the entry, it also
+// performs the post eviction work (clean up for async job and local cache file) for the
+// evicted entry.
+// In case if the cache contains the stale data.FileInfo entry (generation < object.generation)
+// it cleans up (job and local cache file) for the old entry and adds the new entry with the
+// latest generation to the cache.
+//
+// Acquires and releases LOCK(CacheHandler.mu)
+func (chr *CacheHandler) addFileInfoEntryToCache(object *gcs.MinObject, bucket gcs.Bucket) error {
+	fileInfoKey := data.FileInfoKey{
+		BucketName: bucket.Name(),
+		ObjectName: object.Name,
+	}
+	fileInfoKeyName, err := fileInfoKey.Key()
+	if err != nil {
+		return fmt.Errorf("while creating key: %v", fileInfoKeyName)
+	}
+
+	chr.mu.Lock()
+	defer chr.mu.Unlock()
+
+	addEntryToCache := false
+	// Todo - (raj-prince) - this lookUp should not change the LRU order.
+	fileInfo := chr.fileInfoCache.LookUp(fileInfoKeyName)
+	if fileInfo == nil {
+		addEntryToCache = true
+	} else {
+		fileInfoData := fileInfo.(data.FileInfo)
+		if fileInfoData.ObjectGeneration < object.Generation {
+			erasedVal := chr.fileInfoCache.Erase(fileInfoKeyName)
+			if erasedVal != nil {
+				fileInfo := erasedVal.(data.FileInfo)
```

Comment:
let's use the name erasedFileInfo here to avoid confusion with fileInfo used outside if condition.